Compatibility with vector-0.13

This will apparently require code changes:

```

src/System/IO/Streams/Vector.hs:80:46: error:
    • Couldn't match type ‘GHC.ST.ST VM.RealWorld’ with ‘IO’
      Expected type: V.Mutable v VM.RealWorld a -> IO (v a)
        Actual type: V.Mutable v VM.RealWorld a
                     -> GHC.ST.ST VM.RealWorld (v a)
    • In the second argument of ‘(>=>)’, namely ‘basicUnsafeFreeze’
      In the expression: toMutableVectorSized n >=> basicUnsafeFreeze
      In an equation for ‘toVectorSized’:
          toVectorSized n = toMutableVectorSized n >=> basicUnsafeFreeze
   |
80 | toVectorSized n = toMutableVectorSized n >=> V.basicUnsafeFreeze
   |                                              ^^^^^^^^^^^^^^^^^^^

src/System/IO/Streams/Vector.hs:139:5: error:
    • Couldn't match type ‘GHC.ST.ST s0 (v0 a0)’ with ‘IO (v c)’
      Expected type: IO (OutputStream c, IO (v c))
        Actual type: IO (OutputStream c, GHC.ST.ST s0 (v0 a0))
    • In a stmt of a 'do' block:
        return $! (os, flush >>= basicUnsafeFreeze)
      In the expression:
        do (os, flush) <- mutableVectorOutputStreamSized n
           return $! (os, flush >>= basicUnsafeFreeze)
      In an equation for ‘vectorOutputStreamSized’:
          vectorOutputStreamSized n
            = do (os, flush) <- mutableVectorOutputStreamSized n
                 return $! (os, flush >>= basicUnsafeFreeze)
    • Relevant bindings include
        flush :: GHC.ST.ST s0 (V.Mutable v0 s0 a0)
          (bound at src/System/IO/Streams/Vector.hs:138:10)
        os :: OutputStream c
          (bound at src/System/IO/Streams/Vector.hs:138:6)
        vectorOutputStreamSized :: Int -> IO (OutputStream c, IO (v c))
          (bound at src/System/IO/Streams/Vector.hs:137:1)
    |
139 |     return $! (os, flush >>= V.basicUnsafeFreeze)
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/System/IO/Streams/Vector.hs:294:58: error:
    • Couldn't match type ‘GHC.ST.ST VM.RealWorld’ with ‘IO’
      Expected type: V.Mutable v VM.RealWorld a -> IO (v a)
        Actual type: V.Mutable v VM.RealWorld a
                     -> GHC.ST.ST VM.RealWorld (v a)
    • In the second argument of ‘(>=>)’, namely ‘basicUnsafeFreeze’
      In the expression:
        outputToMutableVectorSized n >=> basicUnsafeFreeze
      In an equation for ‘outputToVectorSized’:
          outputToVectorSized n
            = outputToMutableVectorSized n >=> basicUnsafeFreeze
    |
294 | outputToVectorSized n = outputToMutableVectorSized n >=> V.basicUnsafeFreeze
    |                                                          ^^^^^^^^^^^^^^^^^^^
cabal: Failed to build io-streams-1.5.2.1
```

Context: https://github.com/commercialhaskell/stackage/issues/6624

--- a/io-streams.cabal
+++ b/io-streams.cabal
@@ -147,7 +147,7 @@ Library
                      text               >=0.10   && <1.3  || >= 2.0 && <2.1,
                      time               >= 1.2   && <1.13,
                      transformers       >= 0.2   && <0.7,
-                     vector             >= 0.7   && <0.13
+                     vector             >= 0.7   && <0.14
 
   if !impl(ghc >= 7.8)
     Build-depends:   bytestring-builder >= 0.10  && <0.11

--- a/src/System/IO/Streams/Vector.hs
+++ b/src/System/IO/Streams/Vector.hs
@@ -1,6 +1,7 @@
 {-# LANGUAGE BangPatterns          #-}
 {-# LANGUAGE MultiParamTypeClasses #-}
 {-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE CPP                   #-}
 
 -- | Vector conversions and utilities.
 
@@ -29,7 +30,7 @@ module System.IO.Streams.Vector
 import           Control.Concurrent.MVar     (modifyMVar, modifyMVar_, newMVar)
 import           Control.Monad               (liftM, (>=>))
 import           Control.Monad.IO.Class      (MonadIO (..))
-import           Control.Monad.Primitive     (PrimState (..))
+import           Control.Monad.Primitive     (PrimState (..), RealWorld)
 import           Data.IORef                  (IORef, newIORef, readIORef, writeIORef)
 import           Data.Vector.Generic         (Vector (..))
 import qualified Data.Vector.Generic         as V
@@ -38,6 +39,16 @@ import qualified Data.Vector.Generic.Mutable as VM
 import           System.IO.Streams.Internal  (InputStream, OutputStream, fromGenerator, yield)
 import qualified System.IO.Streams.Internal  as S
 
+#if MIN_VERSION_vector(0,13,0)
+import Control.Monad.ST (stToIO)
+#endif
+
+basicUnsafeFreezeCompat :: Vector v a => V.Mutable v RealWorld a -> IO (v a)
+#if MIN_VERSION_vector(0,13,0)
+basicUnsafeFreezeCompat = stToIO . V.basicUnsafeFreeze
+#else
+basicUnsafeFreezeCompat = V.basicUnsafeFreeze
+#endif
 
 ------------------------------------------------------------------------------
 -- | Transforms a vector into an 'InputStream' that yields each of the values
@@ -77,7 +88,7 @@ toVector = toVectorSized dEFAULT_BUFSIZ
 -- | Like 'toVector', but allows control over how large the vector buffer is to
 -- start with.
 toVectorSized :: Vector v a => Int -> InputStream a -> IO (v a)
-toVectorSized n = toMutableVectorSized n >=> V.basicUnsafeFreeze
+toVectorSized n = toMutableVectorSized n >=> basicUnsafeFreezeCompat
 {-# INLINE toVectorSized #-}
 
 
@@ -136,7 +147,7 @@ vectorOutputStream = vectorOutputStreamSized dEFAULT_BUFSIZ
 vectorOutputStreamSized :: Vector v c => Int -> IO (OutputStream c, IO (v c))
 vectorOutputStreamSized n = do
     (os, flush) <- mutableVectorOutputStreamSized n
-    return $! (os, flush >>= V.basicUnsafeFreeze)
+    return $! (os, flush >>= basicUnsafeFreezeCompat)
 
 
 ------------------------------------------------------------------------------
@@ -291,7 +302,7 @@ outputToVectorSized :: Vector v a =>
                        Int
                     -> (OutputStream a -> IO b)
                     -> IO (v a)
-outputToVectorSized n = outputToMutableVectorSized n >=> V.basicUnsafeFreeze
+outputToVectorSized n = outputToMutableVectorSized n >=> basicUnsafeFreezeCompat
 {-# INLINE outputToVectorSized #-}
 
 

